cw_decred/lib/wallet.dart

```diff
@@ -65,63 +82,318 @@ abstract class DecredWalletBase extends WalletBase<DecredBalance,
+    final send = () async {
+      this.checkBalance = true;
+      libdcrwallet.sendRawTransaction(walletInfo.name, signedHex);
+    };
     return DecredPendingTransaction(
-        txid:
-            "3cbf3eb9523fd04e96dbaf98cdbd21779222cc8855ece8700494662ae7578e02",
-        amount: 12345678,
-        fee: 1234,
-        rawHex: "baadbeef");
+        txid: decoded["txid"] ?? "",
+        amount: totalAmt,
+        fee: decoded["fee"] ?? 0,
+        rawHex: signedHex,
+        send: send);
   }
 
   int feeRate(TransactionPriority priority) {
-    // TODO
-    return 1000;
+    final defaultFeeRate = 1000;
+    if (!(priority is DecredTransactionPriority)) {
+      return defaultFeeRate;
+    }
+    int Function(int nb) feeForNb = (int nb) {
+      try {
+        final feeStr = libdcrwallet.estimateFee(walletInfo.name, nb);
+        return int.parse(feeStr);
+      } catch (e) {
+        return defaultFeeRate;
+      }
+    };
+    final p = priority as DecredTransactionPriority;
+    switch (p) {
+      case DecredTransactionPriority.slow:
+        if (feeRateSlow.isOld()) {
+          feeRateSlow.update(feeForNb(4));
+        }
+        return feeRateSlow.fee();
+        break;
+      case DecredTransactionPriority.medium:
+        if (feeRateMedium.isOld()) {
+          feeRateMedium.update(feeForNb(2));
+        }
+        return feeRateMedium.fee();
+        break;
+      case DecredTransactionPriority.fast:
+        if (feeRateFast.isOld()) {
+          feeRateFast.update(feeForNb(1));
+        }
+        return feeRateFast.fee();
+        break;
+    }
+    return defaultFeeRate;
   }
 
   @override
   int calculateEstimatedFee(TransactionPriority priority, int? amount) {
     if (priority is DecredTransactionPriority) {
-      return libdcrwallet.calculateEstimatedFeeWithFeeRate(
-          this.feeRate(priority), amount ?? 0);
-    }
+      final P2PKHOutputSize =
+          36; // 8 bytes value + 2 bytes version + at least 1 byte varint script size + P2PKHPkScriptSize
+      // MsgTxOverhead is 4 bytes version (lower 2 bytes for the real transaction
+      // version and upper 2 bytes for the serialization type) + 4 bytes locktime
+      // + 4 bytes expiry + 3 bytes of varints for the number of transaction
+      // inputs (x2 for witness and prefix) and outputs
+      final MsgTxOverhead = 15;
+      // TxInOverhead is the overhead for a wire.TxIn with a scriptSig length <
+      // 254. prefix (41 bytes) + ValueIn (8 bytes) + BlockHeight (4 bytes) +
+      // BlockIndex (4 bytes) + sig script var int (at least 1 byte)
+      final TxInOverhead = 57;
+      final P2PKHInputSize = TxInOverhead +
+          109; // TxInOverhead (57) + var int (1) + P2PKHSigScriptSize (108)
 
+      // Estimate using a transaction consuming three inputs and paying to one
+      // address with change.
+      return this.feeRate(priority) *
+          (MsgTxOverhead + P2PKHOutputSize * 3 + P2PKHInputSize * 2);
```

Comment:
Oh yeah code is wrong, mean for two outputs.

I'll look at what the other coins are doing again. I don't know if I ever checked.

Reply:
Yeah I think it already knows what it is going to send, but we don't have the whole picture without a lot more code. We are leaving it up to the wallet to pick inputs mostly. Because this is just an estimation, and the user can see the actual fee before sending, I think it's fine if this is quick and dirty. Should we do more here?

I swapped the 2 and 3, so three inputs two outputs.